.forbids pragma: defining forbidden tags

Lancer11211 opened the pull request:
Hi,
I've made this to handle my use case we've discussed under [this forum post](https://forum.nim-lang.org/t/9280). My tests are located at tests/effects/teffects1(1-7).nim. Let me know if you're willing to accept this PR and if anything needs to be done before merging. Thank you!

Araq commented:
Excellent work. But we need to bikeshed about the name. "notTags"? How about "prevents"? "avoids"? "forbids"?

Lancer11211 commented:
Thank you! `forbids` sounds good to me. Let me know if it's final and I'll update the code.

Lancer11211 commented:
Meanwhile, I've added another test: tests/effects/teffects12.nim - this is the one I mentioned in the forum post. It's a simple code which enables safe locking for a resource while disabling recurrent locks.

Commit:
.forbids pragma: defining illegal effects for proc types

This patch intends to define the opposite of the .tags pragma: a way to define effects which are not allowed in a proc.

Lancer11211 commented:
I've renamed the pragma to `forbids` and truncated the commits.

Araq commented:
We also need a manual update, a changelog entry and an RFC. Maybe the RFC got already written though. And I don't say this to demotivate you, I'll help you push it through.

We need to focus especially on the subtyping relation between `type T = proc () {.forbids: [E].}` and `type T = proc () < no forbids specified>`.

Lancer11211 commented:
Of course, I'll update the doc/manual.md and the changelog.md. I'll let you know when it's ready. I think this RFC is relevant: https://github.com/nim-lang/RFCs/issues/302

Commit:
updated documentation and changelogs for the forbids pragma

Commit:
renamed notTagEffects to forbiddenEffects

Commit:
corrected issues of forbids pragma

the forbids pragma didn't handle simple restrictions properly and it also had issues with subtyping

Lancer11211 commented:
While I was writing the documentation, I've found 2 bugs in the patch so I fixed them(and added 2 tests as coverage) before pushing [this documentation/changelog update](https://github.com/nim-lang/Nim/pull/20050/commits/6961b555d6e1a1fbf771072232f1068fd6535ee6) - let me know if you need anything added.

Commit:
removed incorrect character from changelog

Commit:
added test to cover the interaction of methods and the forbids pragma

Commit:
covering the interaction of the tags and forbids pragmas

Araq commented:
The type relation must be documented and justified.

Commit:
updated manual about the forbids pragma

Lancer11211 commented:
Do you mean [like this](https://github.com/nim-lang/Nim/pull/20050/commits/139b4d724ff75211e92a0c6b328207a6dd3e5f4f)?

arnetheduck commented:
> Excellent work. But we need to bikeshed about the name. "notTags"? How about "prevents"? "avoids"? "forbids"?

~~At the core of this need for a new pragma lies a semantic difference: `tags` are additive while `raises` are not, ergo tags needs two pragmas - if we treat tags like exceptions (more or less), the alternative design here is that we have a pragma that works like `raises` which explicitly lists all tags that a function can propagate instead of "adding" and "forbidding".~~

wrong, per commend below

Araq commented on doc/manual.md:
```diff
@@ -4973,6 +4973,15 @@ There is also a way which can be used to forbid certain effects:
+  type ProcType2 = proc (i: int)
+
+Calling `ProcType2` in place of `ProcType1` is valid because `ProcType2` doesn't support any tag, and thus it doesn't allow `MyEffect` either.
+
```
Well no, ProcType2 actually allows *any* effect, it's unspecified aka "don't care".

Lancer11211 replied:
I've changed/corrected the subtyping: https://github.com/nim-lang/Nim/blob/3fc7865c1b706a71cb13662d7e72d32599c61ca8/tests/effects/teffects19.nim
I'm using the same test in the manual and I've tried to explain the subtyping relation.

Araq commented:
>  tags are additive while raises are not

I don't know why you keep phrasing it this way. Raises are additive too, except that `try except` can undo a raises effect. A general "undo tag" feature can be currently approximated via `{.cast(tags: []).}`. With the possible extension `.cast(onlyAllows: [E])`

arnetheduck commented:
> I don't know why you keep phrasing it this way.

my bad, got confused by some test code

Commit:
removed useless statement

Commit:
corrected the subtyping of proc types using the forbids pragma

Commit:
updated manual for the forbids pragma

Lancer11211 commented:
I've corrected the patch to properly support proc type subtyping and I've updated the manual too.

Araq commented on doc/manual.md:
```diff
@@ -4955,6 +4955,52 @@ also be attached to a proc type. This affects type compatibility.
+  caller2(toBeCalled2)
+
+`ProcType2` is a subtype of `ProcType1`. Unlike with tags, the parent context
+doesn't inherit the forbidden list of effects.
```
What does that mean "the parent context doesn't inherit"? What is the "parent context"?

Lancer11211 replied:
Let's say function1 calls function2 - then function1 is the "parent context". I'm calling it a context because it's a context of effects in terms of the effect system. I can't really articulate it, but I wanted to refer to a hierarchy of function calls and statements.

Lancer11211 replied:
I've added a little clarification to that line.

Commit:
updated documentations for forbids pragma

Araq commented:
The CI is red because you need to update the docgen tests too.

Please run:

```
nim c -r -d:nimTestsNimdocFixup nimdoc/tester.nim
```

Commit:
updated nim docs

Commit:
Merge github.com:nim-lang/Nim into devel

Lancer11211 commented:
Thanks! I updated the docs and pull from devel.

Commit:
updated docs with rsttester.nim

Lancer11211 commented:
It seems like I'm still missing something:

```
FAILURE: files differ: nimdoc/testproject/htmldocs/testproject.html
diff --git a/nimdoc/testproject/expected/testproject.html b/nimdoc/testproject/htmldocs/testproject.html
```

I tried `nim c -r -d:nimTestsNimdocFixup nimdoc/testproject/testproject.nim` but it didn't change anything.

Commit:
regenerated documentation

Commit:
updated rst docs

Lancer11211 commented:
Seems like the CI passed. Previously, I used `nim c -r ...` instead of `./bin/nim c -r ...` which messed up the generated files.

Commit:
Merge github.com:nim-lang/Nim into devel

Lancer11211 commented:
Do we need anything else for the merge?

ringabout commented on changelog.md:
```diff
@@ -67,6 +67,8 @@ becomes an alias for `addr`.
 ## Language changes
 
+- [Tag tracking](manual.html#tag-tracking) supports the definition of forbidden tags by the `.forbids` pragma
+  it can be used to disable certain effects in proc types
```
```suggestion
  it can be used to disable certain effects in proc types.
```

Lancer11211 committed:
Update changelog.md

Co-authored-by: ringabout <43030857+ringabout@users.noreply.github.com>

Commit:
updated changelog

jd-yolabs commented on compiler/semcall.nim:
```diff
@@ -154,6 +154,8 @@ proc effectProblem(f, a: PType; result: var string; c: PContext) =
       of efEffectsDelayed:
         result.add "\n  The `.effectsOf` annotations differ."
+      of efTagsIllegal:
+        result.add "\n  The `.forbids` requirements catched an illegal tag."
```
catched -> caught

linkmonitor commented on compiler/semcall.nim:
```diff
@@ -154,6 +154,8 @@ proc effectProblem(f, a: PType; result: var string; c: PContext) =
       of efEffectsDelayed:
         result.add "\n  The `.effectsOf` annotations differ."
+      of efTagsIllegal:
+        result.add "\n  The `.forbids` requirements catched an illegal tag."
```
catched -> caught

Lancer11211 replied:
resolved here: https://github.com/nim-lang/Nim/pull/20050/commits/f9b8f4e36d8cdb6e3eb36d0f260f74a2259a93ec

Commit:
corrected typo

Lancer11211 commented:
Thanks for the review and the merge!

dom96 commented:
This is beautiful, thank you for implementing it!
